NDEV-3821 Change core API to core RPC

artem-yazkov opened the pull request.

artem-yazkov committed:
NODE-3821 new class CoreRpcClient

artem-yazkov committed:
NDEV-3821 wrap up neon-core-rpc binary within CoreRpcServer class

artem-yazkov committed:
NDEV-3821 use CoreRpcClient, CoreRpcServer classes within Proxy App and Indexer App

artem-yazkov committed:
NODE-3821 new config option CORE_API_LIBDIR (/spl/lib)

NODE-3821 change CORE_API_SERVER_BIN value to neon-core-rpc

artem-yazkov committed:
NODE-3821 improve logic for CoreApiServer, CoreRpcServer

artem-yazkov committed:
NODE-3821 autoconnect for CoreRpcClient

artem-yazkov committed:
NODE-3821 common/neon_rpc/client.py: code cleanup

neonlabstech commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSignatureStatuses              |     134 |      0     |      0.001 |    0.000291045 |        0      |
| getBalance                        |     128 |      0     |      0.001 |    0.000265625 |        0      |
| getLatestBlockhash                |      46 |      0     |      0.001 |    0.000217391 |        0      |
| isBlockhashValid                  |      44 |      0     |      0.001 |    9.09091e-05 |        0      |
| requestAirdrop                    |      42 |      0     |      0.001 |    0.000428571 |        0      |
| getSlot                           |      20 |      0     |      0.003 |    0.00055     |        0      |
| getAccountInfo                    |      16 |      0     |      0.008 |    0.00075     |        0      |
| getBlock                          |      16 |      0.001 |      0.012 |    0.0109375   |        0.012  |
| getVersion                        |       8 |      0     |      0.001 |    0.000375    |        0      |
| getTokenAccountBalance            |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| sendTransaction                   |       2 |      0.001 |      0.001 |    0.001       |        0.001  |
| getMinimumBalanceForRentExemption |       1 |      0     |      0     |    0           |        0      |
| getProgramAccounts                |       1 |      0.024 |      0.024 |    0.024       |        0.024  |

</details>

artem-yazkov committed:
NDEV-3821 put Neon RPC client and Neon library into docker image

artem-yazkov committed:
NDEV-3821 fix typo in RPC serialize name: emulate_mutiple -> emulate_multiple

artem-yazkov committed:
Merge branch 'develop' into NDEV-3821

afalaleev commented on common/config/config.py:
```diff
@@ -636,7 +637,7 @@ def neon_core_api_port(self) -> int:
     @cached_property
     def external_neon_core_api(self) -> bool:
-        return self.neon_core_api_ip_name in os.environ
+        return self.neon_core_api_ip not in ["127.0.0.1", "localhost"]
```
The base idea is another.
If the parameter is defined - the Proxy should connect to the defined address, and it doesn't matter on which address the RPC/CORE-API service is listening

artem-yazkov replied:
that's clear, reverted

afalaleev commented on common/neon_rpc/client.py:
```diff
@@ -61,119 +65,245 @@
-    _wait_sec: Final[float] = max(ONE_BLOCK_SEC / 5, 0.05)
-
-    def __init__(self, cfg: Config, sol_client: SolClient, stat_client: RpcStatClient) -> None:
+class CoreRpcRequest(BaseModel):
```
The file should contain only the client's implementation. 
All API structures should be defined in the api.py file.

artem-yazkov replied:
that' clear, reverted

afalaleev commented on common/neon_rpc/client.py:
```diff
@@ -61,119 +65,245 @@
         self._sol_client = sol_client
-
         self._deployed_slot = -1
-        self._token_list_cache: list[TokenModel] = list()
```
You have removed this variable, but it is used bellow

artem-yazkov replied:
reverted

afalaleev commented on common/neon_rpc/client.py:
```diff
@@ -61,119 +65,245 @@
     async def get_evm_cfg(self) -> EvmConfigModel | None:
         try:
-            exec_addr = await self._get_evm_exec_addr()
+            # Load the BPF program account to get the address of the BPF executable account
```
What was the reason to refactor this part of the code?

artem-yazkov replied:
reverted

afalaleev commented on common/neon_rpc/client.py:
```diff
@@ -61,119 +65,245 @@
-        self._token_list_cache: list[TokenModel] = list()
 
-        self._raise_for_status = False
+    async def get_core_api_version(self) -> str:
```
What was the reason to move the method in the source code?

artem-yazkov replied:
reverted

afalaleev commented on common/neon_rpc/client.py:
```diff
@@ -61,119 +65,245 @@
-                extra=self._msg_filter,
-            )
+        try:
+            req = GetHolderRequest.from_dict({"pubkey": address.to_string()})
```
Please refactor to use the Pydantic definitions. Don't use raw dicts.

afalaleev commented on common/neon_rpc/client.py:
```diff
@@ -61,119 +65,245 @@
+            srl_addr_list = []
+            for addr in address_list:
+                srl_addr_list.append({"address": addr.to_address(), "chain_id": addr.chain_id})
+            req = GetBalanceRequest.from_dict({"account": srl_addr_list, "slot": self._get_slot(block)})
```
Please refactor to use the Pydantic definitions, don't use raw dicts.

afalaleev commented on common/neon_rpc/client.py:
```diff
@@ -183,14 +313,31 @@ async def get_state_tx_cnt(self, address: NeonAddress, block: NeonBlockHdrModel
-        req = NeonContractRequest(contract=address.eth_address, slot=self._get_slot(block))
-        resp: CoreApiResp = await self._send_request("contract", req)
-        return NeonContractModel.from_dict(resp.value[0], address=address)
+        req = GetContractRequest.from_dict({"contract": address.to_address(), "slot": self._get_slot(block)})
```
Please use the Pydantic definitions, don't use the raw dicts.

afalaleev commented on common/neon_rpc/client.py:
```diff
@@ -370,123 +519,41 @@ async def emulate_sol_tx_list(
-        )
-        for alt_error in alt_error_list:
-            if sub_error.startswith(alt_error):
-                raise SolAltError("Simulation error: " + alt_error)
```
This list of errors is critical for the correct work of the Proxy in production.
How is it handled by neon-rpc?

artem-yazkov replied:
this logic recreated within `_rpc_error_handler`

afalaleev commented on common/neon_rpc/server.py:
```diff
@@ -115,17 +117,50 @@ def _run_host_api(self, cmd_line: list[str], env: dict[str, Any]):
+class _RpcInstance(_BaseInstance):
+    def __init__(self, cfg: Config, idx: int):
+        super().__init__(cfg, idx)
+        self._run_cmd = [cfg.neon_core_api_server_bin, cfg.neon_core_api_server_libdir]
```
It isn't clear. 
To which SOLANA_URL must connect RPC instance?

artem-yazkov replied:
Sure, I am fix that

artem-yazkov committed:
NDEV-3821 fix field name for GetConfigRespons: NEON_TREE_ACCOUNT_FINISH_TRANSACTION_GAS -> NEON_TREE_ACCOUNT_FINISH_TRANSACTION_FEE

neonlabstech commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   40411 |      0     |      0.25  |    0.000482146 |        0      |
| getBlocks                         |   28114 |      0     |      0.233 |    0.000313082 |        0      |
| getBlock                          |   14210 |      0     |      0.036 |    0.000573962 |        0.001  |
| getTransaction                    |   11994 |      0     |      0.231 |    0.00173603  |        0.001  |
| getMultipleAccounts               |    8881 |      0     |      0.254 |    0.0013056   |        0.001  |
| sendTransaction                   |    5914 |      0     |      0.25  |    0.00516233  |        0.002  |
| getLatestBlockhash                |    5402 |      0     |      0.074 |    0.000343206 |        0      |
| getBalance                        |    1910 |      0     |      0.03  |    0.000272775 |        0      |
| getSignatureStatuses              |     479 |      0     |      0.06  |    0.000507307 |        0      |
| isBlockhashValid                  |     217 |      0     |      0.001 |    0.000198157 |        0      |
| getVersion                        |     183 |      0     |      0.011 |    0.000382514 |        0      |
| getSignaturesForAddress           |      99 |      0     |      0.002 |    0.000878788 |        0.001  |
| simulateTransaction               |      42 |      0     |      0.003 |    0.000309524 |        0      |
| requestAirdrop                    |      41 |      0     |      0.001 |    0.000585366 |        0.001  |
| getBlockHeight                    |      27 |      0     |      0.001 |    3.7037e-05  |        0      |
| getSlot                           |      25 |      0     |      0.001 |    0.0004      |        0      |
| getTokenAccountBalance            |       3 |      0     |      0     |    0           |        0      |
| getFirstAvailableBlock            |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getHealth                         |       2 |      0     |      0     |    0           |        0      |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |        0.001  |
| getProgramAccounts                |       1 |      0.041 |      0.041 |    0.041       |        0.041  |
| getClusterNodes                   |       1 |      0     |      0     |    0           |        0      |

</details>

artem-yazkov committed:
NDEV-3821 all RPC calls use exists API from common.neon_rpc.api

artem-yazkov committed:
NDEV-3821 allow serialize RPC responses directly to NeonAccountModel, NeonContractModel, EvmConfigModel, HolderAccountModel

artem-yazkov committed:
NDEV-3821 catch interface changes of HolderAccountModel class

artem-yazkov committed:
NDEV-3821 remove debug and dead code

vodolaz commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/16515408812/artifacts/3613059082)

vodolaz commented:
<details><summary><strong>⚠️ Cost Report Validation Failed ⚠️</strong></summary>

| dapp           | action                    | metric   |   INCREASE |
|:---------------|:--------------------------|:---------|-----------:|
| aave           | liquidation               | gas_used |      12600 |
| saddle finance | add liquidity in metapool | gas_used |      14700 |

artem-yazkov committed:
NDEV-3821 HolderAccountModel.from_cls(), NeonAccountModel.from_cls() factories

artem-yazkov committed:
NDEV-3821 CoreRpcClient minor refactoring

artem-yazkov committed:
NDEV-3821 redefine external_neon_core_api: it's true when CORE_API_IP envar defined

vodolaz commented:
<details><summary><strong>⚠️ Cost Report Validation Failed ⚠️</strong></summary>

| dapp   | action                   | metric    |   INCREASE |
|:-------|:-------------------------|:----------|-----------:|
| aave   | borrow from lending pool | gas_used  |      18900 |
| aave   | liquidation              | gas_used  |      12600 |
| aave   | repay                    | trx_count |          3 |
| aave   | repay                    | gas_used  |     190050 |

vodolaz commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   40573 |      0     |      0.262 |    0.000481157 |         0     |
| getBlocks                         |   27702 |      0     |      0.243 |    0.000300375 |         0     |
| getBlock                          |   13990 |      0     |      0.11  |    0.000541959 |         0.001 |
| getTransaction                    |   12164 |      0     |      0.137 |    0.00237373  |         0.001 |
| getMultipleAccounts               |    8912 |      0     |      0.265 |    0.00153523  |         0.001 |
| sendTransaction                   |    5992 |      0     |      0.256 |    0.00564269  |         0.002 |
| getLatestBlockhash                |    5436 |      0     |      0.051 |    0.000409492 |         0     |
| getBalance                        |    1918 |      0     |      0.043 |    0.000281543 |         0     |
| getSignatureStatuses              |     478 |      0     |      0.141 |    0.000748954 |         0     |
| isBlockhashValid                  |     216 |      0     |      0.001 |    0.000203704 |         0     |
| getVersion                        |     182 |      0     |      0.007 |    0.000313187 |         0     |
| getSignaturesForAddress           |     105 |      0     |      0.002 |    0.000952381 |         0.001 |
| requestAirdrop                    |      42 |      0     |      0.001 |    0.00047619  |         0     |
| simulateTransaction               |      42 |      0     |      0.005 |    0.000809524 |         0     |
| getSlot                           |      25 |      0     |      0.002 |    0.00052     |         0     |
| getBlockHeight                    |      23 |      0     |      0.008 |    0.000652174 |         0     |
| getTokenAccountBalance            |       3 |      0     |      0     |    0           |         0     |
| getFirstAvailableBlock            |       2 |      0     |      0     |    0           |         0     |
| getHealth                         |       2 |      0     |      0     |    0           |         0     |
| getMinimumBalanceForRentExemption |       1 |      0     |      0     |    0           |         0     |
| getProgramAccounts                |       1 |      0.033 |      0.033 |    0.033       |         0.033 |
| getClusterNodes                   |       1 |      0     |      0     |    0           |         0     |

</details>

artem-yazkov committed:
NDEV-3821 class CoreRpcServer: take SOLANA_URL from cfg.sol_url_list

vodolaz commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/16521173779/artifacts/3615220908)

vodolaz commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   40224 |      0     |      0.23  |    0.000436481 |        0      |
| getBlocks                         |   26917 |      0     |      0.201 |    0.00030516  |        0      |
| getBlock                          |   13614 |      0     |      0.227 |    0.000589246 |        0.001  |
| getTransaction                    |   12007 |      0     |      0.062 |    0.00177921  |        0.001  |
| getMultipleAccounts               |    8834 |      0     |      0.267 |    0.00134163  |        0.001  |
| sendTransaction                   |    5944 |      0     |      0.23  |    0.00557301  |        0.002  |
| getLatestBlockhash                |    5380 |      0     |      0.117 |    0.00037342  |        0      |
| getBalance                        |    1899 |      0     |      0.008 |    0.000238547 |        0      |
| getSignatureStatuses              |     466 |      0     |      0.002 |    0.000287554 |        0      |
| isBlockhashValid                  |     210 |      0     |      0.002 |    0.000195238 |        0      |
| getVersion                        |     176 |      0     |      0.002 |    0.000238636 |        0      |
| getSignaturesForAddress           |      98 |      0     |      0.002 |    0.00094898  |        0.001  |
| requestAirdrop                    |      42 |      0     |      0.001 |    0.000380952 |        0      |
| simulateTransaction               |      42 |      0     |      0.001 |    0.000357143 |        0      |
| getSlot                           |      25 |      0     |      0.002 |    0.00052     |        0      |
| getBlockHeight                    |      25 |      0     |      0.001 |    0.00024     |        0      |
| getTokenAccountBalance            |       3 |      0     |      0.001 |    0.000333333 |        0      |
| getFirstAvailableBlock            |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getHealth                         |       2 |      0.001 |      0.001 |    0.001       |        0.001  |
| getMinimumBalanceForRentExemption |       1 |      0     |      0     |    0           |        0      |
| getProgramAccounts                |       1 |      0.034 |      0.034 |    0.034       |        0.034  |
| getClusterNodes                   |       1 |      0     |      0     |    0           |        0      |

</details>

artem-yazkov committed:
NDEV-3821 use union CoreServer class; pass ApiInstance, RpcInstance into constructor

artem-yazkov committed:
NDEV-3821 common/neon_rpc/client.py: reduce import names

artem-yazkov committed:
NDEV-3821 add exception_handler method, name and _wait_sec variables to the CoreRpcClient class

afalaleev commented on common/neon_rpc/server.py:
```diff
@@ -74,23 +74,25 @@ def _create_env(self) -> dict[str, Any]:
             SOLANA_RAYON_THREADS="1",
         )
+        if self._solana_url:
+            new_env["SOLANA_URL"]=self._solana_url
```
This change looks unnecessary

artem-yazkov replied:
fixed

afalaleev commented on common/neon_rpc/client.py:
```diff
@@ -1,9 +1,13 @@
 import itertools
 import logging
-from typing import Sequence, Final, TypeVar, ClassVar, Union
+import uuid
```
These modules aren't used:
- unittest
- pprint
- asyncio
- uuid
- pydantic

artem-yazkov replied:
unused modules were removed

afalaleev commented on common/neon_rpc/client.py:
```diff
@@ -184,13 +151,22 @@ async def get_state_tx_cnt(self, address: NeonAddress, block: NeonBlockHdrModel
-        resp: CoreApiResp = await self._send_request("contract", req)
-        return NeonContractModel.from_dict(resp.value[0], address=address)
+        resp = await self._get_contract(req)
+        return NeonContractModel(neon_address=address, code=resp[0].code, solana_address=resp[0].sol_address)
```
NeonContractModel has the `sol_address` property.

artem-yazkov replied:
fixed

artem-yazkov committed:
NDEV-3821 disable assert on CoreRpcClient::_exception_handler

artem-yazkov committed:
Revert "NDEV-3821 use union CoreServer class; pass ApiInstance, RpcInstance into constructor"

This reverts commit 94b51f05cf123b01f769baa481f4e7eb998f599c.

artem-yazkov committed:
Revert "NDEV-3821 class CoreRpcServer: take SOLANA_URL from cfg.sol_url_list"

This reverts commit 1d80aee14a10f05981004818f4348c1756c43a3a.

artem-yazkov committed:
NDEV-3821 disable CoreRpcClient._exception_handler

vodolaz commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/16570019142/artifacts/3629790206)

vodolaz commented:
<details><summary><strong>⚠️ Cost Report Validation Failed ⚠️</strong></summary>

| dapp       | action                                          | metric        |   INCREASE |
|:-----------|:------------------------------------------------|:--------------|-----------:|
| aave       | borrow from lending pool                        | gas_used      |      16800 |
| aave       | liquidation                                     | gas_used      |      10500 |
| uniswap v3 | nonfungiblepositionmanager - decrease liquidity | acc_count     |          1 |
| uniswap v3 | nonfungiblepositionmanager - decrease liquidity | gas_estimated |     983040 |
| uniswap v3 | nonfungiblepositionmanager - decrease liquidity | gas_used      |     934740 |
| uniswap v3 | nonfungiblepositionmanager - mint position      | trx_count     |          3 |
| uniswap v3 | nonfungiblepositionmanager - mint position      | gas_used      |      75800 |

vodolaz commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   40459 |       0    |      0.24  |    0.00045095  |        0      |
| getBlocks                         |   28460 |       0    |      0.2   |    0.000303127 |        0      |
| getBlock                          |   14389 |       0    |      0.23  |    0.000555772 |        0      |
| getTransaction                    |   12094 |       0    |      0.22  |    0.00211948  |        0.001  |
| getMultipleAccounts               |    8873 |       0    |      0.091 |    0.0012431   |        0.001  |
| sendTransaction                   |    5958 |       0    |      0.243 |    0.00540232  |        0.002  |
| getLatestBlockhash                |    5399 |       0    |      0.195 |    0.000403593 |        0      |
| getBalance                        |    1908 |       0    |      0.012 |    0.00025891  |        0      |
| getSignatureStatuses              |     486 |       0    |      0.212 |    0.000751029 |        0      |
| isBlockhashValid                  |     220 |       0    |      0.001 |    0.000154545 |        0      |
| getVersion                        |     186 |       0    |      0.001 |    0.000268817 |        0      |
| getSignaturesForAddress           |     104 |       0    |      0.002 |    0.000961538 |        0.001  |
| requestAirdrop                    |      42 |       0    |      0.002 |    0.000380952 |        0      |
| simulateTransaction               |      42 |       0    |      0.001 |    0.000380952 |        0      |
| getBlockHeight                    |      28 |       0    |      0.001 |    0.00025     |        0      |
| getSlot                           |      25 |       0    |      0.001 |    0.0004      |        0      |
| getTokenAccountBalance            |       3 |       0    |      0.001 |    0.000333333 |        0      |
| getFirstAvailableBlock            |       2 |       0    |      0.001 |    0.0005      |        0.0005 |
| getHealth                         |       2 |       0    |      0.001 |    0.0005      |        0.0005 |
| getMinimumBalanceForRentExemption |       1 |       0    |      0     |    0           |        0      |
| getProgramAccounts                |       1 |       0.03 |      0.03  |    0.03        |        0.03   |
| getClusterNodes                   |       1 |       0    |      0     |    0           |        0      |

</details>

artem-yazkov committed:
NDEV-3821 add _rpc_error_handler method to JsonRpcClient, CoreRpcClient

artem-yazkov committed:
NDEV-3821 united CoreServer for RPC and API protocols

artem-yazkov committed:
NDEV-3821 remove dead code

NDEV-3821 fix issue with NeonContractModel.sol_address

artem-yazkov committed:
NDEV-3821 optimize server environment creation

vodolaz commented:
<details><summary><strong>⚠️ Cost Report Validation Failed ⚠️</strong></summary>

| dapp       | action                                     | metric    |   INCREASE |
|:-----------|:-------------------------------------------|:----------|-----------:|
| aave       | borrow from lending pool                   | acc_count |          1 |
| aave       | deposit to lending pool                    | acc_count |          1 |
| aave       | deposit to lending pool                    | trx_count |          2 |
| aave       | deposit to lending pool                    | gas_used  |     114100 |
| aave       | repay                                      | trx_count |          3 |
| aave       | repay                                      | gas_used  |     127500 |
| aave       | withdraw                                   | acc_count |          1 |
| uniswap v3 | nonfungiblepositionmanager - mint position | gas_used  |      14700 |

vodolaz commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   40185 |      0     |      0.238 |    0.000449098 |        0      |
| getBlocks                         |   26824 |      0     |      0.193 |    0.000307933 |        0      |
| getBlock                          |   13572 |      0     |      0.067 |    0.000559092 |        0.001  |
| getTransaction                    |   12034 |      0     |      0.133 |    0.00240211  |        0.001  |
| getMultipleAccounts               |    8813 |      0     |      0.241 |    0.00136821  |        0.001  |
| sendTransaction                   |    5928 |      0     |      0.238 |    0.00537567  |        0.002  |
| getLatestBlockhash                |    5377 |      0     |      0.06  |    0.00032639  |        0      |
| getBalance                        |    1897 |      0     |      0.027 |    0.000248814 |        0      |
| getSignatureStatuses              |     482 |      0     |      0.002 |    0.000259336 |        0      |
| isBlockhashValid                  |     218 |      0     |      0.001 |    0.000252294 |        0      |
| getVersion                        |     184 |      0     |      0.006 |    0.000331522 |        0      |
| getSignaturesForAddress           |     107 |      0     |      0.002 |    0.000953271 |        0.001  |
| requestAirdrop                    |      42 |      0     |      0.001 |    0.000380952 |        0      |
| simulateTransaction               |      42 |      0     |      0.001 |    0.000285714 |        0      |
| getBlockHeight                    |      28 |      0     |      0.002 |    0.000285714 |        0      |
| getSlot                           |      25 |      0     |      0.001 |    0.00044     |        0      |
| getTokenAccountBalance            |       3 |      0     |      0     |    0           |        0      |
| getFirstAvailableBlock            |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getHealth                         |       2 |      0     |      0     |    0           |        0      |
| getMinimumBalanceForRentExemption |       1 |      0     |      0     |    0           |        0      |
| getProgramAccounts                |       1 |      0.042 |      0.042 |    0.042       |        0.042  |
| getClusterNodes                   |       1 |      0.001 |      0.001 |    0.001       |        0.001  |

</details>

artem-yazkov committed:
NDEV-3821 log RPC Error Handler output

vodolaz commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/16615969614/artifacts/3647047772)

artem-yazkov committed:
NDEV-3821 Use RpcClientRequest instead of HttpClientRequest within JsonRpcClient

vodolaz commented:
<details><summary><strong>⚠️ Cost Report Validation Failed ⚠️</strong></summary>

| dapp       | action                                     | metric    |   INCREASE |
|:-----------|:-------------------------------------------|:----------|-----------:|
| aave       | liquidation                                | acc_count |          1 |
| aave       | liquidation                                | gas_used  |      10500 |
| aave       | withdraw                                   | trx_count |          2 |
| aave       | withdraw                                   | gas_used  |     116200 |
| uniswap v3 | nonfungiblepositionmanager - mint position | gas_used  |      10500 |

vodolaz commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   40262 |      0     |      0.234 |    0.000469227 |         0     |
| getBlocks                         |   27920 |      0     |      0.218 |    0.000306196 |         0     |
| getBlock                          |   14122 |      0     |      0.225 |    0.000561464 |         0.001 |
| getTransaction                    |   11977 |      0     |      0.157 |    0.00197712  |         0.001 |
| getMultipleAccounts               |    8821 |      0     |      0.24  |    0.0013893   |         0.001 |
| sendTransaction                   |    5898 |      0     |      0.237 |    0.00546592  |         0.002 |
| getLatestBlockhash                |    5377 |      0     |      0.198 |    0.000412126 |         0     |
| getBalance                        |    1902 |      0     |      0.063 |    0.000293375 |         0     |
| getSignatureStatuses              |     480 |      0     |      0.044 |    0.000433333 |         0     |
| isBlockhashValid                  |     217 |      0     |      0.002 |    0.000207373 |         0     |
| getVersion                        |     183 |      0     |      0.003 |    0.00026776  |         0     |
| getSignaturesForAddress           |     106 |      0     |      0.003 |    0.000896226 |         0.001 |
| requestAirdrop                    |      42 |      0     |      0.002 |    0.000547619 |         0.001 |
| simulateTransaction               |      42 |      0     |      0.001 |    0.000333333 |         0     |
| getBlockHeight                    |      29 |      0     |      0.008 |    0.00062069  |         0     |
| getSlot                           |      25 |      0     |      0.001 |    0.00028     |         0     |
| getTokenAccountBalance            |       3 |      0     |      0     |    0           |         0     |
| getFirstAvailableBlock            |       2 |      0     |      0     |    0           |         0     |
| getHealth                         |       2 |      0.001 |      0.001 |    0.001       |         0.001 |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |         0.001 |
| getProgramAccounts                |       1 |      0.033 |      0.033 |    0.033       |         0.033 |
| getClusterNodes                   |       1 |      0.001 |      0.001 |    0.001       |         0.001 |

</details>

artem-yazkov committed:
NDEV-3821 make _stat_client of RpcStatInfo optional

artem-yazkov committed:
NDEV-3821 refactor JsonRpcClient method callback

artem-yazkov committed:
NDEV-3821 revert CoreRpcClient Exception Handler

artem-yazkov committed:
NDEV-3821 improve _register_batch_sender callback

artem-yazkov committed:
NDEV-3821 remove assert for method.is_async_def in _register_batch_sender()

vodolaz commented:
<details><summary><strong>⚠️ Cost Report Validation Failed ⚠️</strong></summary>

| dapp           | action                     | metric    |   INCREASE |
|:---------------|:---------------------------|:----------|-----------:|
| aave           | borrow from lending pool   | acc_count |          1 |
| aave           | deposit to lending pool    | acc_count |          1 |
| aave           | liquidation                | acc_count |          1 |
| aave           | liquidation                | gas_used  |      12600 |
| aave           | repay                      | acc_count |          1 |
| aave           | withdraw                   | acc_count |          1 |
| saddle finance | swap usdc -> susd metapool | acc_count |          1 |

artem-yazkov committed:
NDEV-3821 avoid endless loops in JsonRpcClient senders

artem-yazkov committed:
NDEV-3821 passtrough exceptions from _extract_return()

artem-yazkov committed:
NDEV-3821 restore original JsonRpcClient senders

NDEV-3821 don't log _rpc_error_handler() error message, just pass it to stat server

vodolaz commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/16650220765/artifacts/3658636302)

vodolaz commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   40221 |      0     |      0.244 |    0.000466199 |         0     |
| getBlocks                         |   26886 |      0     |      0.22  |    0.00030462  |         0     |
| getBlock                          |   13604 |      0     |      0.034 |    0.000528006 |         0     |
| getTransaction                    |   11886 |      0     |      0.217 |    0.00180616  |         0.001 |
| getMultipleAccounts               |    8794 |      0     |      0.25  |    0.00133784  |         0.001 |
| sendTransaction                   |    5860 |      0     |      0.244 |    0.00540154  |         0.002 |
| getLatestBlockhash                |    5350 |      0     |      0.071 |    0.000339252 |         0     |
| getBalance                        |    1902 |      0     |      0.036 |    0.000293375 |         0     |
| getSignatureStatuses              |     479 |      0     |      0.006 |    0.000304802 |         0     |
| isBlockhashValid                  |     217 |      0     |      0.001 |    0.000202765 |         0     |
| getVersion                        |     184 |      0     |      0.004 |    0.000228261 |         0     |
| getSignaturesForAddress           |      98 |      0     |      0.002 |    0.000938776 |         0.001 |
| simulateTransaction               |      42 |      0     |      0.001 |    0.000404762 |         0     |
| requestAirdrop                    |      41 |      0     |      0.001 |    0.000292683 |         0     |
| getSlot                           |      25 |      0     |      0.001 |    0.00048     |         0     |
| getBlockHeight                    |      25 |      0     |      0.001 |    0.00016     |         0     |
| getTokenAccountBalance            |       3 |      0     |      0.001 |    0.000333333 |         0     |
| getFirstAvailableBlock            |       2 |      0.001 |      0.001 |    0.001       |         0.001 |
| getHealth                         |       2 |      0     |      0     |    0           |         0     |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |         0.001 |
| getProgramAccounts                |       1 |      0.029 |      0.029 |    0.029       |         0.029 |
| getClusterNodes                   |       1 |      0     |      0     |    0           |         0     |

</details>

artem-yazkov committed:
NDEV-3821 error_list can't be None

artem-yazkov committed:
NDEV-3821 improve JsonRpcClient single sender

vodolaz commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/16670717049/artifacts/3666008607)

vodolaz commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   40168 |      0     |      0.231 |    0.000470997 |        0      |
| getBlocks                         |   26856 |      0     |      0.152 |    0.000307194 |        0      |
| getBlock                          |   13587 |      0     |      0.08  |    0.000558696 |        0.001  |
| getTransaction                    |   11832 |      0     |      0.2   |    0.00223174  |        0.001  |
| getMultipleAccounts               |    8774 |      0     |      0.2   |    0.00130989  |        0.001  |
| sendTransaction                   |    5834 |      0     |      0.232 |    0.00530614  |        0.002  |
| getLatestBlockhash                |    5345 |      0     |      0.205 |    0.000364266 |        0      |
| getBalance                        |    1900 |      0     |      0.024 |    0.000244211 |        0      |
| getSignatureStatuses              |     469 |      0     |      0.051 |    0.000490405 |        0      |
| isBlockhashValid                  |     212 |      0     |      0.001 |    0.000174528 |        0      |
| getVersion                        |     179 |      0     |      0.011 |    0.000312849 |        0      |
| getSignaturesForAddress           |     102 |      0     |      0.003 |    0.00105882  |        0.001  |
| simulateTransaction               |      42 |      0     |      0.001 |    0.000404762 |        0      |
| requestAirdrop                    |      41 |      0     |      0.001 |    0.000439024 |        0      |
| getBlockHeight                    |      29 |      0     |      0.001 |    0.000241379 |        0      |
| getSlot                           |      25 |      0     |      0.002 |    0.0008      |        0.001  |
| getTokenAccountBalance            |       3 |      0     |      0     |    0           |        0      |
| getFirstAvailableBlock            |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getHealth                         |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |        0.001  |
| getProgramAccounts                |       1 |      0.038 |      0.038 |    0.038       |        0.038  |
| getClusterNodes                   |       1 |      0.001 |      0.001 |    0.001       |        0.001  |

</details>

artem-yazkov committed:
NDEV-3821 improve JsonRpcClient batch sender

artem-yazkov committed:
NDEV-3821 improve JsonRpcClient batch sender. Bugfix

artem-yazkov committed:
NDEV-3821 improve JsonRpcClient batch sender. Bugfix 2

vodolaz commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/16673869015/artifacts/3666984099)

vodolaz commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/16673869015/artifacts/3667682808)

vodolaz commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   40279 |      0     |      0.289 |    0.00048025  |        0      |
| getBlocks                         |   27097 |      0     |      0.196 |    0.000304794 |        0      |
| getBlock                          |   13711 |      0     |      0.287 |    0.000561666 |        0.001  |
| getTransaction                    |   12023 |      0     |      0.075 |    0.00219654  |        0.001  |
| getMultipleAccounts               |    8831 |      0     |      0.287 |    0.00130257  |        0.001  |
| sendTransaction                   |    5965 |      0     |      0.288 |    0.00564208  |        0.002  |
| getLatestBlockhash                |    5373 |      0     |      0.031 |    0.000332961 |        0      |
| getBalance                        |    1901 |      0     |      0.006 |    0.000251973 |        0      |
| getSignatureStatuses              |     482 |      0     |      0.006 |    0.000342324 |        0      |
| isBlockhashValid                  |     218 |      0     |      0.001 |    0.000183486 |        0      |
| getVersion                        |     184 |      0     |      0.004 |    0.000336957 |        0      |
| getSignaturesForAddress           |     107 |      0     |      0.003 |    0.00107477  |        0.001  |
| requestAirdrop                    |      42 |      0     |      0.001 |    0.000547619 |        0.001  |
| simulateTransaction               |      42 |      0     |      0.002 |    0.000380952 |        0      |
| getSlot                           |      25 |      0     |      0.001 |    0.0004      |        0      |
| getBlockHeight                    |      25 |      0     |      0.006 |    0.00044     |        0      |
| getTokenAccountBalance            |       3 |      0     |      0     |    0           |        0      |
| getFirstAvailableBlock            |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getHealth                         |       2 |      0     |      0     |    0           |        0      |
| getMinimumBalanceForRentExemption |       1 |      0     |      0     |    0           |        0      |
| getProgramAccounts                |       1 |      0.038 |      0.038 |    0.038       |        0.038  |
| getClusterNodes                   |       1 |      0     |      0     |    0           |        0      |

</details>

artem-yazkov committed:
Merge branch 'develop' into NDEV-3821

vodolaz commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/16677754862/artifacts/3668238075)

vodolaz commented:
<details><summary><strong>⚠️ Cost Report Validation Failed ⚠️</strong></summary>

| dapp           | action                                          | metric    |   INCREASE |
|:---------------|:------------------------------------------------|:----------|-----------:|
| aave           | borrow from lending pool                        | acc_count |          1 |
| aave           | liquidation                                     | gas_used  |      18900 |
| aave           | repay                                           | acc_count |          1 |
| saddle finance | swap dai -> usdt metapool                       | gas_used  |      14700 |
| uniswap v3     | nonfungiblepositionmanager - increase liquidity | gas_used  |      21000 |
| uniswap v3     | nonfungiblepositionmanager - mint position      | gas_used  |      18900 |

vodolaz commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   40666 |      0     |      0.239 |    0.000450106 |        0      |
| getBlocks                         |   27333 |      0     |      0.213 |    0.000312845 |        0      |
| getBlock                          |   13823 |      0     |      0.04  |    0.000519931 |        0      |
| getTransaction                    |   12248 |      0     |      0.079 |    0.0018483   |        0.001  |
| getMultipleAccounts               |    8941 |      0     |      0.242 |    0.0013522   |        0.001  |
| sendTransaction                   |    6065 |      0     |      0.239 |    0.00566645  |        0.002  |
| getLatestBlockhash                |    5444 |      0     |      0.078 |    0.000373622 |        0      |
| getBalance                        |    1926 |      0     |      0.006 |    0.000239356 |        0      |
| getSignatureStatuses              |     492 |      0     |      0.133 |    0.000630081 |        0      |
| isBlockhashValid                  |     223 |      0     |      0.002 |    0.000188341 |        0      |
| getVersion                        |     189 |      0     |      0.002 |    0.000190476 |        0      |
| getSignaturesForAddress           |     100 |      0     |      0.003 |    0.00107     |        0.001  |
| requestAirdrop                    |      42 |      0     |      0.001 |    0.000404762 |        0      |
| simulateTransaction               |      42 |      0     |      0.001 |    0.000428571 |        0      |
| getSlot                           |      25 |      0     |      0.001 |    0.00044     |        0      |
| getBlockHeight                    |      25 |      0     |      0.002 |    0.00016     |        0      |
| getTokenAccountBalance            |       3 |      0     |      0.001 |    0.000333333 |        0      |
| getFirstAvailableBlock            |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getHealth                         |       2 |      0     |      0     |    0           |        0      |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |        0.001  |
| getProgramAccounts                |       1 |      0.026 |      0.026 |    0.026       |        0.026  |
| getClusterNodes                   |       1 |      0     |      0     |    0           |        0      |

</details>
